services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/base_scheduler.py

```diff
@@ -0,0 +1,339 @@
+from ...models.domains.comp_pipelines import CompPipelineAtDB
+from ...models.domains.comp_runs import CompRunsAtDB
+from ...models.domains.comp_tasks import CompTaskAtDB
+from ...models.schemas.constants import UserID
+from ...utils.computations import get_pipeline_state_from_task_states
+from ...utils.scheduler import COMPLETED_STATES, Iteration, get_repository
+from ..db.repositories.comp_pipelines import CompPipelinesRepository
+from ..db.repositories.comp_runs import CompRunsRepository
+from ..db.repositories.comp_tasks import CompTasksRepository
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ScheduledPipelineParams:
+    mark_for_cancellation: bool = False
+
+
+@dataclass
+class BaseCompScheduler(ABC):
+    scheduled_pipelines: Dict[
+        Tuple[UserID, ProjectID, Iteration], ScheduledPipelineParams
+    ]
+    db_engine: Engine
+    wake_up_event: asyncio.Event = field(default_factory=asyncio.Event, init=False)
+
+    async def run_new_pipeline(self, user_id: UserID, project_id: ProjectID) -> None:
+        # ensure the pipeline exists and is populated with something
+        dag = await self._get_pipeline_dag(project_id)
+        if not dag:
+            logger.warning(
+                "project %s has no computational dag defined. not scheduled for a run."
+            )
+            return
+
+        runs_repo: CompRunsRepository = get_repository(
+            self.db_engine, CompRunsRepository
+        )  # type: ignore
+        new_run: CompRunsAtDB = await runs_repo.create(
+            user_id=user_id, project_id=project_id
+        )
+        self.scheduled_pipelines[
+            (user_id, project_id, new_run.iteration)
+        ] = ScheduledPipelineParams()
+        # ensure the scheduler starts right away
+        self._wake_up_scheduler_now()
+
+    async def stop_pipeline(
+        self, user_id: UserID, project_id: ProjectID, iteration: Optional[int] = None
+    ) -> None:
+        if not iteration:
+            # if no iteration given find the latest one in the list
+            possible_iterations = {
+                it
+                for u_id, p_id, it in self.scheduled_pipelines
+                if u_id == user_id and p_id == project_id
+            }
+            if not possible_iterations:
+                raise SchedulerError(
+                    f"There are no pipeline scheduled for {user_id}:{project_id}"
+                )
+            iteration = max(possible_iterations)
+
+        # mark the scheduled pipeline for stopping
+        self.scheduled_pipelines[
+            (user_id, project_id, iteration)
+        ].mark_for_cancellation = True
+        # ensure the scheduler starts right away
+        self._wake_up_scheduler_now()
+
+    async def schedule_all_pipelines(self) -> None:
+        self.wake_up_event.clear()
+        # if one of the task throws, the other are NOT cancelled which is what we want
+        await asyncio.gather(
```

Comment:
should we limit the tasks passed to gather to an upper limit?

Reply:
that might be a thing. but I'll keep it for later.